diff -u a/ChainLattice.c b/ChainLattice.c
--- a/ChainLattice.c
+++ b/ChainLattice.c
@@ -96,7 +96,7 @@
     StdFace_PrintVal_d("D", &StdI->D[2][2], 0.0);
     StdFace_InputSpinNN(StdI->J, StdI->JAll, StdI->J0, StdI->J0All, "J0");
     StdFace_InputSpinNN(StdI->Jp, StdI->JpAll, StdI->J0p, StdI->J0pAll, "J0'");
-    StdFace_InputSpinNN(StdI->Jpp, StdI->JppAll, StdI->J0pp, StdI->J0ppAll, "J0'");
+    StdFace_InputSpinNN(StdI->Jpp, StdI->JppAll, StdI->J0pp, StdI->J0ppAll, "J0''");
     /**/
     StdFace_NotUsed_d("mu", StdI->mu);
     StdFace_NotUsed_d("U", StdI->U);
Only in a: CMakeLists.txt
diff -u a/HoneycombLattice.c b/HoneycombLattice.c
--- a/HoneycombLattice.c
+++ b/HoneycombLattice.c
@@ -82,9 +82,9 @@
     StdFace_InputSpinNN(StdI->Jp, StdI->JpAll, StdI->J0p, StdI->J0pAll, "J0'");
     StdFace_InputSpinNN(StdI->Jp, StdI->JpAll, StdI->J1p, StdI->J1pAll, "J1'");
     StdFace_InputSpinNN(StdI->Jp, StdI->JpAll, StdI->J2p, StdI->J2pAll, "J2'");
-    StdFace_InputSpinNN(StdI->Jpp, StdI->JppAll, StdI->J0pp, StdI->J0ppAll, "J0'");
-    StdFace_InputSpinNN(StdI->Jpp, StdI->JppAll, StdI->J1pp, StdI->J1ppAll, "J1'");
-    StdFace_InputSpinNN(StdI->Jpp, StdI->JppAll, StdI->J2pp, StdI->J2ppAll, "J2'");
+    StdFace_InputSpinNN(StdI->Jpp, StdI->JppAll, StdI->J0pp, StdI->J0ppAll, "J0''");
+    StdFace_InputSpinNN(StdI->Jpp, StdI->JppAll, StdI->J1pp, StdI->J1ppAll, "J1''");
+    StdFace_InputSpinNN(StdI->Jpp, StdI->JppAll, StdI->J2pp, StdI->J2ppAll, "J2''");
     /**/
     StdFace_NotUsed_d("mu", StdI->mu);
     StdFace_NotUsed_d("U", StdI->U);
@@ -96,6 +96,10 @@
     StdFace_NotUsed_c("t0'", StdI->t0p);
     StdFace_NotUsed_c("t1'", StdI->t1p);
     StdFace_NotUsed_c("t2'", StdI->t2p);
+    StdFace_NotUsed_c("t''", StdI->tpp);
+    StdFace_NotUsed_c("t0''", StdI->t0pp);
+    StdFace_NotUsed_c("t1''", StdI->t1pp);
+    StdFace_NotUsed_c("t2''", StdI->t2pp);
     StdFace_NotUsed_d("V", StdI->V);
     StdFace_NotUsed_d("V0", StdI->V0);
     StdFace_NotUsed_d("V1", StdI->V1);
@@ -104,6 +108,10 @@
     StdFace_NotUsed_d("V0'", StdI->V0p);
     StdFace_NotUsed_d("V1'", StdI->V1p);
     StdFace_NotUsed_d("V2'", StdI->V2p);
+    StdFace_NotUsed_d("V''", StdI->Vpp);
+    StdFace_NotUsed_d("V0''", StdI->V0pp);
+    StdFace_NotUsed_d("V1''", StdI->V1pp);
+    StdFace_NotUsed_d("V2''", StdI->V2pp);
   }/*if (strcmp(StdI->model, "spin") == 0 )*/
   else {
     StdFace_PrintVal_d("mu", &StdI->mu, 0.0);
@@ -126,12 +134,18 @@
     StdFace_InputCoulombV(StdI->Vpp, &StdI->V0pp, "V0''");
     StdFace_InputCoulombV(StdI->Vpp, &StdI->V1pp, "V1''");
     StdFace_InputCoulombV(StdI->Vpp, &StdI->V2pp, "V2''");
-    StdFace_PrintVal_d("V'", &StdI->Vp, 0.0);
     /**/
     StdFace_NotUsed_J("J0", StdI->J0All, StdI->J0);
     StdFace_NotUsed_J("J1", StdI->J1All, StdI->J1);
     StdFace_NotUsed_J("J2", StdI->J2All, StdI->J2);
     StdFace_NotUsed_J("J'", StdI->JpAll, StdI->Jp);
+    StdFace_NotUsed_J("J0'", StdI->J0pAll, StdI->J0p);
+    StdFace_NotUsed_J("J1'", StdI->J1pAll, StdI->J1p);
+    StdFace_NotUsed_J("J2'", StdI->J2pAll, StdI->J2p);
+    StdFace_NotUsed_J("J''", StdI->JppAll, StdI->Jpp);
+    StdFace_NotUsed_J("J0''", StdI->J0ppAll, StdI->J0pp);
+    StdFace_NotUsed_J("J1''", StdI->J1ppAll, StdI->J1pp);
+    StdFace_NotUsed_J("J2''", StdI->J2ppAll, StdI->J2pp);
     StdFace_NotUsed_d("D", StdI->D[2][2]);
 
     if (strcmp(StdI->model, "hubbard") == 0 ) {
diff -u a/Kagome.c b/Kagome.c
--- a/Kagome.c
+++ b/Kagome.c
@@ -90,7 +90,6 @@
     StdFace_NotUsed_d("mu", StdI->mu);
     StdFace_NotUsed_d("U", StdI->U);
     StdFace_NotUsed_c("t", StdI->t);
-    StdFace_NotUsed_c("t0", StdI->t);
     StdFace_NotUsed_c("t0", StdI->t0);
     StdFace_NotUsed_c("t1", StdI->t1);
     StdFace_NotUsed_c("t2", StdI->t2);
@@ -106,6 +105,18 @@
     StdFace_NotUsed_d("V0'", StdI->V0p);
     StdFace_NotUsed_d("V1'", StdI->V1p);
     StdFace_NotUsed_d("V2'", StdI->V2p);
+    StdFace_NotUsed_c("t''", StdI->tpp);
+    StdFace_NotUsed_c("t0''", StdI->t0pp);
+    StdFace_NotUsed_c("t1''", StdI->t1pp);
+    StdFace_NotUsed_c("t2''", StdI->t2pp);
+    StdFace_NotUsed_d("V''", StdI->Vpp);
+    StdFace_NotUsed_d("V0''", StdI->V0pp);
+    StdFace_NotUsed_d("V1''", StdI->V1pp);
+    StdFace_NotUsed_d("V2''", StdI->V2pp);
+    StdFace_NotUsed_J("J''", StdI->JppAll, StdI->Jpp);
+    StdFace_NotUsed_J("J0''", StdI->J0ppAll, StdI->J0pp);
+    StdFace_NotUsed_J("J1''", StdI->J1ppAll, StdI->J1pp);
+    StdFace_NotUsed_J("J2''", StdI->J2ppAll, StdI->J2pp);
   }/*if (strcmp(StdI->model, "spin") == 0 )*/
   else {
     StdFace_PrintVal_d("mu", &StdI->mu, 0.0);
@@ -130,6 +141,18 @@
     StdFace_NotUsed_J("J0'", StdI->J0pAll, StdI->J0p);
     StdFace_NotUsed_J("J1'", StdI->J1pAll, StdI->J1p);
     StdFace_NotUsed_J("J2'", StdI->J2pAll, StdI->J2p);
+    StdFace_NotUsed_J("J''", StdI->JppAll, StdI->Jpp);
+    StdFace_NotUsed_J("J0''", StdI->J0ppAll, StdI->J0pp);
+    StdFace_NotUsed_J("J1''", StdI->J1ppAll, StdI->J1pp);
+    StdFace_NotUsed_J("J2''", StdI->J2ppAll, StdI->J2pp);
+    StdFace_NotUsed_c("t''", StdI->tpp);
+    StdFace_NotUsed_c("t0''", StdI->t0pp);
+    StdFace_NotUsed_c("t1''", StdI->t1pp);
+    StdFace_NotUsed_c("t2''", StdI->t2pp);
+    StdFace_NotUsed_d("V''", StdI->Vpp);
+    StdFace_NotUsed_d("V0''", StdI->V0pp);
+    StdFace_NotUsed_d("V1''", StdI->V1pp);
+    StdFace_NotUsed_d("V2''", StdI->V2pp);
     StdFace_NotUsed_d("D", StdI->D[2][2]);
 
     if (strcmp(StdI->model, "hubbard") == 0 ) {
diff -u a/Ladder.c b/Ladder.c
--- a/Ladder.c
+++ b/Ladder.c
@@ -62,8 +62,13 @@
   StdFace_RequiredVal_i("W", StdI->W);
   StdFace_NotUsed_i("a0W", StdI->box[0][0]);
   StdFace_NotUsed_i("a0L", StdI->box[0][1]);
+  StdFace_NotUsed_i("a0H", StdI->box[0][2]);
   StdFace_NotUsed_i("a1W", StdI->box[1][0]);
   StdFace_NotUsed_i("a1L", StdI->box[1][1]);
+  StdFace_NotUsed_i("a1H", StdI->box[1][2]);
+  StdFace_NotUsed_i("a2W", StdI->box[2][0]);
+  StdFace_NotUsed_i("a2L", StdI->box[2][1]);
+  StdFace_NotUsed_i("a2H", StdI->box[2][2]);
   /**/
   StdFace_PrintVal_d("phase0", &StdI->phase[0], 0.0);
   StdFace_NotUsed_d("phase1", StdI->phase[1]);
@@ -72,7 +77,8 @@
   /**/
   StdI->NsiteUC = StdI->W;
   StdI->W = 1;
-  StdI->direct[0][0] = (double)StdI->NsiteUC;
+  StdI->direct[0][0] *= (double)StdI->NsiteUC;
+  StdI->direct[0][1] *= (double)StdI->NsiteUC;
   StdFace_InitSite(StdI, fp, 2);
   for (isite = 0; isite < StdI->NsiteUC; isite++){
     StdI->tau[isite][0] = (double)isite / (double)StdI->NsiteUC;
@@ -82,11 +88,8 @@
   (2) check & store parameters of Hamiltonian
   */
   fprintf(stdout, "\n  @ Hamiltonian \n\n");
-  StdFace_NotUsed_J("J", StdI->JAll, StdI->J);
   StdFace_NotUsed_J("J'", StdI->JpAll, StdI->Jp);
-  StdFace_NotUsed_c("t", StdI->t);
   StdFace_NotUsed_c("t'", StdI->tp);
-  StdFace_NotUsed_d("V", StdI->V);
   StdFace_NotUsed_d("V'", StdI->Vp);
   StdFace_NotUsed_d("K", StdI->K);
   StdFace_PrintVal_d("h", &StdI->h, 0.0);
@@ -102,6 +105,9 @@
     StdFace_InputSpin(StdI->J1p, StdI->J1pAll, "J1'");
     StdFace_InputSpin(StdI->J2p, StdI->J2pAll, "J2'");
     /**/
+    StdFace_NotUsed_J("J", StdI->JAll, StdI->J);
+    StdFace_NotUsed_c("t", StdI->t);
+    StdFace_NotUsed_d("V", StdI->V);
     StdFace_NotUsed_d("mu", StdI->mu);
     StdFace_NotUsed_d("U", StdI->U);
     StdFace_NotUsed_c("t0", StdI->t0);
@@ -132,8 +138,8 @@
     StdFace_NotUsed_J("J0", StdI->J0All, StdI->J0);
     StdFace_NotUsed_J("J1", StdI->J1All, StdI->J1);
     StdFace_NotUsed_J("J2", StdI->J2All, StdI->J2);
-    StdFace_NotUsed_J("J1p", StdI->J1pAll, StdI->J1p);
-    StdFace_NotUsed_J("J2p", StdI->J2pAll, StdI->J2p);
+    StdFace_NotUsed_J("J1'", StdI->J1pAll, StdI->J1p);
+    StdFace_NotUsed_J("J2'", StdI->J2pAll, StdI->J2p);
     StdFace_NotUsed_d("D", StdI->D[2][2]);
 
     if (strcmp(StdI->model, "hubbard") == 0 ) {
Only in a: makefile_StdFace
diff -u a/SquareLattice.c b/SquareLattice.c
--- a/SquareLattice.c
+++ b/SquareLattice.c
@@ -124,7 +124,6 @@
     StdFace_InputCoulombV(StdI->Vp, &StdI->V1p, "V1'");
     StdFace_InputCoulombV(StdI->Vpp, &StdI->V0pp, "V0''");
     StdFace_InputCoulombV(StdI->Vpp, &StdI->V1pp, "V1''");
-    StdFace_PrintVal_d("V'", &StdI->Vp, 0.0);
     /**/
     StdFace_NotUsed_J("J0", StdI->J0All, StdI->J0);
     StdFace_NotUsed_J("J1", StdI->J1All, StdI->J1);
diff -u a/StdFace_ModelUtil.c b/StdFace_ModelUtil.c
--- a/StdFace_ModelUtil.c
+++ b/StdFace_ModelUtil.c
@@ -1276,6 +1276,7 @@
   /**@brief
   (1) Transfer StdIntList::trans, StdIntList::transindx
   */
+  ntransMax *= 2;/*HubbardLocal can append 6 on-site terms per site, not 4*/
   StdI->transindx = (int **)malloc(sizeof(int*) * ntransMax);
   StdI->trans = (double complex *)malloc(sizeof(double complex) * ntransMax);
   for (ii = 0; ii < ntransMax; ii++) {
diff -u a/TriangularLattice.c b/TriangularLattice.c
--- a/TriangularLattice.c
+++ b/TriangularLattice.c
@@ -80,9 +80,9 @@
     StdFace_InputSpinNN(StdI->Jp, StdI->JpAll, StdI->J0p, StdI->J0pAll, "J0'");
     StdFace_InputSpinNN(StdI->Jp, StdI->JpAll, StdI->J1p, StdI->J1pAll, "J1'");
     StdFace_InputSpinNN(StdI->Jp, StdI->JpAll, StdI->J2p, StdI->J2pAll, "J2'");
-    StdFace_InputSpinNN(StdI->Jpp, StdI->JppAll, StdI->J0pp, StdI->J0ppAll, "J0'");
-    StdFace_InputSpinNN(StdI->Jpp, StdI->JppAll, StdI->J1pp, StdI->J1ppAll, "J1'");
-    StdFace_InputSpinNN(StdI->Jpp, StdI->JppAll, StdI->J2pp, StdI->J2ppAll, "J2'");
+    StdFace_InputSpinNN(StdI->Jpp, StdI->JppAll, StdI->J0pp, StdI->J0ppAll, "J0''");
+    StdFace_InputSpinNN(StdI->Jpp, StdI->JppAll, StdI->J1pp, StdI->J1ppAll, "J1''");
+    StdFace_InputSpinNN(StdI->Jpp, StdI->JppAll, StdI->J2pp, StdI->J2ppAll, "J2''");
     /**/
     StdFace_NotUsed_d("mu", StdI->mu);
     StdFace_NotUsed_d("U", StdI->U);
@@ -94,7 +94,7 @@
     StdFace_NotUsed_c("t0'", StdI->t0p);
     StdFace_NotUsed_c("t1'", StdI->t1p);
     StdFace_NotUsed_c("t2'", StdI->t2p);
-    StdFace_NotUsed_c("t''", StdI->tp);
+    StdFace_NotUsed_c("t''", StdI->tpp);
     StdFace_NotUsed_c("t0''", StdI->t0pp);
     StdFace_NotUsed_c("t1''", StdI->t1pp);
     StdFace_NotUsed_c("t2''", StdI->t2pp);
@@ -142,6 +142,8 @@
     StdFace_NotUsed_J("J0''", StdI->J0ppAll, StdI->J0pp);
     StdFace_NotUsed_J("J1''", StdI->J1ppAll, StdI->J1pp);
     StdFace_NotUsed_J("J2''", StdI->J2ppAll, StdI->J2pp);
+    StdFace_NotUsed_J("J'", StdI->JpAll, StdI->Jp);
+    StdFace_NotUsed_J("J''", StdI->JppAll, StdI->Jpp);
     StdFace_NotUsed_d("D", StdI->D[2][2]);
 
     if (strcmp(StdI->model, "hubbard") == 0 ) {
